lib/SPIRVProducerPass.cpp

```diff
@@ -3220,144 +3262,802 @@ spv::Op SPIRVProducerPass::GetSPIRVBinaryOpcode(Instruction &I) {
+
+        Ops.clear();
+        Ops << FixedVectorType::get(Type::getInt32Ty(Context), 4) << RID
+            << CstInt0;
+
+        RID = addSPIRVInst(spv::OpCompositeConstruct, Ops);
+      } else if (dim != components) {
+        // get_image_dim return an int2 regardless of the arrayedness of the
+        // image. If the image is arrayed an element must be dropped from the
+        // query result.
+        //
+
+        // Implement:
+        //   %result = OpVectorShuffle %uint2 %sizes %sizes 0 1
+        Ops.clear();
+        Ops << FixedVectorType::get(Type::getInt32Ty(Context), 2) << RID << RID
+            << 0 << 1;
+
+        RID = addSPIRVInst(spv::OpVectorShuffle, Ops);
+      }
+    } else if (components > 1) {
+      // Implement:
+      //     %result = OpCompositeExtract %uint %sizes <component number>
+      Ops.clear();
+      Ops << Call->getType() << RID;
+
+      uint32_t component = 0;
+      if (FuncInfo.getType() == Builtins::kGetImageHeight)
+        component = 1;
+      else if (FuncInfo.getType() == Builtins::kGetImageDepth)
+        component = 2;
+      Ops << component;
+
+      RID = addSPIRVInst(spv::OpCompositeExtract, Ops);
+    }
+    break;
+  }
+  default:
+    llvm_unreachable("Unsupported Image builtin");
+  }
+
+  return RID;
+}
+
+SPIRVID
+SPIRVProducerPass::GenerateSubgroupInstruction(CallInst *Call,
+                                               const FunctionInfo &FuncInfo) {
+  SPIRVID RID;
+
+  // requires SPIRV version 1.3 or greater
+  if (SPIRVMinorVersion < 3)
+    llvm_unreachable(
+        "Sub-group support requires SPIR-V version 1.3 or greater");
+
+  static const std::map<Builtins::BuiltinType, spv::BuiltIn> clspvBuiltinMap = {
+      {Builtins::kGetSubGroupSize, spv::BuiltInSubgroupSize},
+      {Builtins::kGetMaxSubGroupSize, spv::BuiltInSubgroupMaxSize},
+      {Builtins::kGetNumSubGroups, spv::BuiltInNumSubgroups},
+      {Builtins::kGetSubGroupId, spv::BuiltInSubgroupId},
+      {Builtins::kGetSubGroupLocalId, spv::BuiltInSubgroupLocalInvocationId},
+  };
+
+  auto FuncType = FuncInfo.getType();
+
+  auto MII = clspvBuiltinMap.find(FuncType);
+  if (MII != clspvBuiltinMap.end()) {
+    SPIRVOperandVec Ops;
+    Ops << Call->getType() << getSPIRVBuiltin(MII->second);
+
+    return addSPIRVInst(spv::OpLoad, Ops);
+  }
+
+  spv::Capability capability = spv::CapabilityGroupNonUniformBallot;
+  spv::Op op = spv::OpNop;
+  switch (FuncInfo.getType()) {
+  case Builtins::kGetEnqueuedNumSubGroups:
+    break;
+  case Builtins::kSubGroupBarrier:
+    break;
+
+  case Builtins::kSubGroupAll:
+    capability = spv::CapabilityGroupNonUniformVote;
+    op = spv::OpGroupNonUniformAll;
+    break;
+  case Builtins::kSubGroupAny:
+    capability = spv::CapabilityGroupNonUniformVote;
+    op = spv::OpGroupNonUniformAny;
+    break;
+  case Builtins::kSubGroupBroadcast:
+    capability = spv::CapabilityGroupNonUniformBallot;
+    op = spv::OpGroupNonUniformBroadcast;
```

Comment:
Prior to 1.5 the broadcast id must be a constant value. This deserves a note in the docs (or an error if we limit ourselves to generating SPIR-V 1.3).